File: stacks/dynamic-dashboard/genui-controlled-shiny/app/builders/scatter.py
"""ScatterOutput → Plotly figure. Pure; the card just renders what this returns."""

from __future__ import annotations

from typing import Any

import plotly.graph_objects as go

from .theme import PALETTE, base_layout


def scatter_figure(out: dict[str, Any], title: str | None = None) -> go.Figure:
    spec, points = out["spec"], out["points"]
    groups: dict[str, list[dict[str, Any]]] = {}
    for p in points:
        groups.setdefault(p["color"] or "all", []).append(p)
    has_label = bool(points) and points[0]["label"] is not None
    fig = go.Figure()
    for i, (name, pts) in enumerate(groups.items()):
        fig.add_trace(
            go.Scatter(
                mode="markers",
                name=name,
                x=[p["x"] for p in pts],
                y=[p["y"] for p in pts],
                text=[p["label"] or "" for p in pts],
                hovertemplate=(
                    f"{spec['x']}=%{{x}}<br>{spec['y']}=%{{y}}{'<br>%{text}' if has_label else ''}<extra>{name}</extra>"
                ),
                marker={"color": PALETTE[i % len(PALETTE)], "size": 8, "opacity": 0.8},
            )
        )
    fig.update_layout(
        **base_layout(title or f"{spec['y']} vs {spec['x']}"),
        xaxis={"title": {"text": spec["x"]}, "zeroline": False},
        yaxis={"title": {"text": spec["y"]}, "zeroline": False},
        showlegend=spec.get("color") is not None,
    )
    return fig

File: stacks/dynamic-dashboard/genui-controlled-nextjs/src/builders/scatter.ts
/** ScatterOutput → Plotly figure. Pure; the card just renders what this returns. */
import type { Data, Layout } from "plotly.js";
import type { ScatterOutput } from "@/catalog/output";
import { PALETTE, baseLayout } from "./theme";

export function scatterFigure(out: ScatterOutput): { data: Data[]; layout: Partial<Layout> } {
  const { spec, points } = out;
  const groups = new Map<string, typeof points>();
  for (const p of points) {
    const k = p.color ?? "all";
    const g = groups.get(k) ?? [];
    g.push(p);
    groups.set(k, g);
  }
  const data: Data[] = [...groups.entries()].map(([name, pts], i) => ({
    type: "scatter",
    mode: "markers",
    name,
    x: pts.map((p) => p.x),
    y: pts.map((p) => p.y),
    text: pts.map((p) => p.label ?? ""),
    hovertemplate: `${spec.x}=%{x}<br>${spec.y}=%{y}${pts[0]?.label !== null ? "<br>%{text}" : ""}<extra>${name}</extra>`,
    marker: { color: PALETTE[i % PALETTE.length], size: 8, opacity: 0.8 },
  }));
  const layout: Partial<Layout> = {
    ...baseLayout(`${spec.y} vs ${spec.x}`),
    xaxis: { title: { text: spec.x }, zeroline: false },
    yaxis: { title: { text: spec.y }, zeroline: false },
    showlegend: spec.color !== undefined,
  };
  return { data, layout };
}
